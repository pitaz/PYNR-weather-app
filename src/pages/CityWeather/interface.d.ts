export interface CityWeatherProps {}

export interface Event {
  event: {
    target: {
      value: string
    }
  }
}