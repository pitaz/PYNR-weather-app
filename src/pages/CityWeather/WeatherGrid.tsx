import React, { FC } from "react";
import { Grid } from "@material-ui/core";
import ArrowBackIosRounded from "@material-ui/icons/ArrowBackIosRounded";
import ArrowForwardIosRounded from "@material-ui/icons/ArrowForwardIosRounded";
import WeatherCard from "../../components/WeatherCard";

interface WeatherGridProps {
  page: number;
  pages: number;
  items: [];
  handlePrev: () => void;
  value: any;
  setTime: any;
  setTemp: any;
  handleNext: () => void;
}

const WeatherGrid: FC<WeatherGridProps> = ({
  page,
  pages,
  items,
  handlePrev,
  value,
  setTime,
  setTemp,
  handleNext,
}) => {
  return (
    <Grid item xs={12}>
      <Grid container justifyContent="center" alignItems="center" spacing={2}>
        {page > 0 ? <ArrowBackIosRounded onClick={handlePrev} /> : null}
        {items?.map((item) => (
          <Grid key={item} item>
            <WeatherCard
              data={item}
              unit={value}
              setTime={setTime}
              setTemp={setTemp}
            />
          </Grid>
        ))}
        {page + 1 < pages && <ArrowForwardIosRounded onClick={handleNext} />}
      </Grid>
    </Grid>
  );
};

export default WeatherGrid;
