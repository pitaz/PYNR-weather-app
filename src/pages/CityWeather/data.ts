export const data = {
  labels: ["12 pm", "1 pm", "3 pm", "4 pm", "5 pm", "6 pm"],
  datasets: [
    {
      label: "Temp",
      data: [100, 70, 80, 5, 20, 35],
      backgroundColor: "rgb(255, 99, 132)",
    },
  ],
};

export const options = {
  scales: {
    yAxes: [
      {
        stacked: true,
        ticks: {
          beginAtZero: true,
        },
      },
    ],
    xAxes: [
      {
        stacked: true,
      },
    ],
  },
};