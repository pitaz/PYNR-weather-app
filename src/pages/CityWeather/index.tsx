import React, { Suspense } from 'react';
import loadable from "@loadable/component";

 const WeatherComponent =  loadable(() => import("./CityWeather"));

export default () => {
  return (
    <div>
      <Suspense fallback={<div>Loading...</div>}>
        <WeatherComponent />
      </Suspense>
    </div>
  );
}
