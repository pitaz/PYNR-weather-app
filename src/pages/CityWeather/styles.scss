.city-weather {
  .page-title {
    font-family: 'Roboto';
    font-size: 30px;
    font-weight: 500;
    letter-spacing: 0.5 rem;
    font-style: normal;
  }
}