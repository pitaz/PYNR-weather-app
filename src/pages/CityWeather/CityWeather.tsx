import React, { FC, useEffect, useState } from "react";
import { makeStyles } from "@material-ui/core/styles";
import {
  Container,
  CssBaseline,
  Grid,
  RadioGroup,
  Radio,
} from "@material-ui/core";
import { Bar } from "react-chartjs-2";
import FormControlLabel from "@material-ui/core/FormControlLabel";
import CustomButton from "../../components/Button";
import "./styles.scss";
import { options } from "./data";
import { CityWeatherProps } from "./interface";
import { useDispatch } from "react-redux";
import { fetchWeather } from "../../redux/actions/weather";
import { groupBy } from "../../utils/groupBy";
import { checkDevice } from "../../utils/checkMobile";
import ScreenLoader from "../../components/ScreenLoader";
import { useCallback } from "react";
import WeatherGrid from "./WeatherGrid";

/**
 * This component returns the Weather card and bar chart
 * @returns {JSX}
 */
const CityWeather: FC<CityWeatherProps> = () => {
  const classes = useStyles();
  const [value, setValue] = useState<any>("metric");
  const [list, setList] = useState<any>();
  const [perPage, setPerpage] = useState<number>(3);
  const [page, setPage] = useState<number>(0);
  const [pages, setPages] = useState<number>(0);
  const [temp, setTemp] = useState<any>();
  const [time, setTime] = useState<any>();
  const [error, setError] = useState<any>();
  const [loading, setLoading] = useState<boolean>(false);
  const isMobile = checkDevice();
  const dispatch = useDispatch<any>();

  /**
   * Handles radio event
   * @param event event
   */
  const handleRadioChange = (event: any) => {
    setValue(event.target.value);
  };

  useEffect(() => {
    if (isMobile) {
      setPerpage(1);
    }
  }, [isMobile]);

  const fetch = useCallback(async () => {
    try {
      setLoading(true);
      const res = await dispatch(fetchWeather(value));
      const groupData = groupBy(res?.payload?.data?.list, "dt_txt");
      const reslist = Object.entries(groupData);
      setList(reslist);
      setPages(Math.ceil(reslist?.length / perPage));
      setLoading(false);
    } catch (error) {
      setError(error);
      setLoading(false);
    }
  }, [perPage, value]);

  useEffect(() => {
    fetch();
  }, [value, perPage]);

  const items = list?.slice(page * perPage, (page + 1) * perPage);
  const chartData = {
    labels: time,
    datasets: [
      {
        label: "Temp",
        data: temp,
        backgroundColor: "rgb(255, 99, 132)",
      },
    ],
  };

  if (loading) {
    return <ScreenLoader />;
  }

  if (error) {
    return  <h1 className="page-title">An eror occurred</h1>;
  }
  const handlePrev = () => {
    if (page > 0) {
      setPage(page - 1);
    }
  };

  const handleNext = () => {
    const actualPage = page + 1;
    if (actualPage < pages) {
      setPage(page + 1);
    }
  };
  return (
    <React.Fragment>
      <CssBaseline />
      <Container className="city-weather">
        <h1 className="page-title">Weather app</h1>
        <Grid item xs={12}>
          <Grid
            container
            justifyContent="center"
            alignItems="center"
            spacing={1}
          >
            <RadioGroup
              aria-label="temp"
              name="temp"
              className={classes.radioGroup}
              value={value}
              onChange={handleRadioChange}
            >
              <FormControlLabel
                id="unit-celsius"
                value="metric"
                control={<Radio />}
                label="Celsius"
              />
              <FormControlLabel
                id="unit-fahrenheit"
                value="imperial"
                control={<Radio />}
                label="Fahrenheit"
              />
            </RadioGroup>
            <CustomButton title="refresh" onClick={fetch} />
          </Grid>
        </Grid>
        <WeatherGrid
          page={page}
          pages={pages}
          items={items}
          value={value}
          handlePrev={handlePrev}
          handleNext={handleNext}
          setTemp={setTemp}
          setTime={setTime}
        />
        <Bar data={chartData} options={options} height={120} />
      </Container>
    </React.Fragment>
  );
};

const useStyles = makeStyles((theme) => ({
  card: {
    height: 240,
    width: 204,
  },
  bar: {
    padding: "20px 30px",
  },
  title: {
    fontSize: 14,
  },
  radioGroup: {
    flexDirection: "row",
  },
}));

export default CityWeather;
