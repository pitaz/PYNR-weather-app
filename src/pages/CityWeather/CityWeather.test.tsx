/**
 * @jest-environment jsdom
 */

 import React from "react";
 import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

import CityWeather from ".";

describe('Weather city page', () => {
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });
  const wrapper = shallow(<CityWeather />);
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
