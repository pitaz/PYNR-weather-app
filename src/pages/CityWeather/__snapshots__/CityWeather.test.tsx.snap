// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather city page should match snapshot 1`] = `ShallowWrapper {}`;
